src/Processors/Sources/MongoDBSource.cpp

```diff
@@ -1,465 +1,489 @@
+        {
+            String json = "[";
+            auto first = true;
 
-                assert_cast<ColumnUInt32 &>(column).getData().push_back(
-                    static_cast<UInt32>(static_cast<const Poco::MongoDB::ConcreteElement<Poco::Timestamp> &>(value).value().epochTime()));
-                break;
-            }
-            case ValueType::vtUUID:
+            for (const auto & elem : value.get_array().value)
             {
-                if (value.type() == Poco::MongoDB::ElementTraits<String>::TypeId)
+                if (!first)
+                    json += ',';
+                switch (elem.type())
                 {
-                    String string = static_cast<const Poco::MongoDB::ConcreteElement<String> &>(value).value();
-                    assert_cast<ColumnUUID &>(column).getData().push_back(parse<UUID>(string));
+                    // types which need to be quoted
+                    case bsoncxx::type::k_binary:
+                    case bsoncxx::type::k_date:
+                    case bsoncxx::type::k_timestamp:
+                    case bsoncxx::type::k_oid:
+                    case bsoncxx::type::k_symbol:
+                    case bsoncxx::type::k_maxkey:
+                    case bsoncxx::type::k_minkey:
+                    case bsoncxx::type::k_undefined:
+                        json += "\"" + BSONElementAsString<bsoncxx::array::element>(elem) + "\"";
+                        break;
+                    // types which need to be escaped
+                    case bsoncxx::type::k_string:
+                    case bsoncxx::type::k_code:
+                        json += "\"" + escapeJSONString(BSONElementAsString<bsoncxx::array::element>(elem)) + "\"";
+                        break;
+                    default:
+                        json += BSONElementAsString<bsoncxx::array::element>(elem);
                 }
-                else
-                    throw Exception(ErrorCodes::TYPE_MISMATCH, "Type mismatch, expected String (UUID), got type id = {} for column {}",
-                                        toString(value.type()), name);
-                break;
+                first = false;
             }
-            case ValueType::vtArray:
-            {
-                if (value.type() != Poco::MongoDB::ElementTraits<MongoArray::Ptr>::TypeId)
-                    throw Exception(ErrorCodes::TYPE_MISMATCH, "Type mismatch, expected Array, got type id = {} for column {}",
-                                    toString(value.type()), name);
 
-                size_t expected_dimensions = array_info[idx].num_dimensions;
-                const auto parse_value = array_info[idx].parser;
-                std::vector<Row> dimensions(expected_dimensions + 1);
+            json += ']';
+            return json;
+        }
+        case bsoncxx::type::k_binary:
+            return base64Encode(std::string(reinterpret_cast<const char*>(value.get_binary().bytes), value.get_binary().size));
```

Comment:
Why we encode it with base64? We can store raw data in clickhouse `String`s. 
UPD: But it won't work for the case where binary is inside document or array that we serialize to json. How does Mongo return this data with native client when we request document with binary field?

Reply:
Okay, good idea.

As BSON binary field, the same as when outside document.